File: scriptable/src/color.js
/**
 * color.js — 色計算コア
 *
 * E-Ink 4色パレット定義、RGB→Lab変換、CIEDE2000色差、最近色検索。
 * Python実装 (domain/color.py) の忠実なJSポート。
 */

// --- E-Ink 4色パレット ---

const EINK_PALETTE = [
  [255, 255, 255], // White
  [0, 0, 0],       // Black
  [200, 0, 0],     // Red
  [255, 255, 0],   // Yellow
];

// --- RGB → Lab 変換 ---

/**
 * sRGBコンポーネント(0-255)をリニアRGBに変換。
 * @param {number} c - sRGB値 (0-255)
 * @returns {number} リニアRGB値
 */
function srgbToLinear(c) {
  const v = c / 255.0;
  if (v <= 0.04045) {
    return v / 12.92;
  }
  return Math.pow((v + 0.055) / 1.055, 2.4);
}

/**
 * Lab変換の補助関数。
 * @param {number} t
 * @returns {number}
 */
function _labF(t) {
  const delta = 6.0 / 29.0;
  if (t > delta * delta * delta) {
    return Math.pow(t, 1.0 / 3.0);
  }
  return t / (3.0 * delta * delta) + 4.0 / 29.0;
}

/**
 * Lab逆変換の補助関数（_labFの逆関数）。
 * @param {number} t
 * @returns {number}
 */
function _labFInv(t) {
  const delta = 6.0 / 29.0;
  if (t > delta) {
    return t * t * t;
  }
  return 3.0 * delta * delta * (t - 4.0 / 29.0);
}

/**
 * RGB(0-255)をCIE L*a*b*に変換。D65光源基準。
 * @param {number} r - Red (0-255)
 * @param {number} g - Green (0-255)
 * @param {number} b - Blue (0-255)
 * @returns {number[]} [L, a, b]
 */
function rgbToLab(r, g, b) {
  // sRGB → リニアRGB
  const rLin = srgbToLinear(r);
  const gLin = srgbToLinear(g);
  const bLin = srgbToLinear(b);

  // リニアRGB → XYZ (D65)
  const x = 0.4124564 * rLin + 0.3575761 * gLin + 0.1804375 * bLin;
  const y = 0.2126729 * rLin + 0.7151522 * gLin + 0.0721750 * bLin;
  const z = 0.0193339 * rLin + 0.1191920 * gLin + 0.9503041 * bLin;

  // D65 白色点
  const xn = 0.95047, yn = 1.00000, zn = 1.08883;

  // XYZ → L*a*b*
  const fx = _labF(x / xn);
  const fy = _labF(y / yn);
  const fz = _labF(z / zn);

  const L = 116.0 * fy - 16.0;
  const a = 500.0 * (fx - fy);
  const bStar = 200.0 * (fy - fz);

  return [L, a, bStar];
}

/**
 * CIE L*a*b* をRGB(0-255)に変換。D65光源基準。
 * Python lab_to_rgb_batch (color_space.py) の単ピクセル版。
 * @param {number} L - L* (0-100)
 * @param {number} a - a*
 * @param {number} bStar - b*
 * @returns {number[]} [R, G, B] (0-255, clamped)
 */
function labToRgb(L, a, bStar) {
  // Lab → XYZ
  const fy = (L + 16.0) / 116.0;
  const fx = a / 500.0 + fy;
  const fz = fy - bStar / 200.0;

  const x = 0.95047 * _labFInv(fx);
  const y = 1.00000 * _labFInv(fy);
  const z = 1.08883 * _labFInv(fz);

  // XYZ → リニアRGB
  const rLin =  3.2404542 * x - 1.5371385 * y - 0.4985314 * z;
  const gLin = -0.9692660 * x + 1.8760108 * y + 0.0415560 * z;
  const bLin =  0.0556434 * x - 0.2040259 * y + 1.0572252 * z;

  // リニアRGB → sRGB
  function toSrgb(c) {
    c = Math.max(0, Math.min(1, c));
    if (c <= 0.0031308) return 12.92 * c;
    return 1.055 * Math.pow(c, 1.0 / 2.4) - 0.055;
  }

  return [
    Math.max(0, Math.min(255, Math.round(toSrgb(rLin) * 255))),
    Math.max(0, Math.min(255, Math.round(toSrgb(gLin) * 255))),
    Math.max(0, Math.min(255, Math.round(toSrgb(bLin) * 255))),
  ];
}

/**
 * CIEDE2000色差を計算。
 * 参考: "The CIEDE2000 Color-Difference Formula" (Sharma et al., 2005)
 * @param {number[]} lab1 - [L, a, b]
 * @param {number[]} lab2 - [L, a, b]
 * @returns {number} CIEDE2000色差
 */
function ciede2000(lab1, lab2) {
  const [l1, a1, b1] = lab1;
  const [l2, a2, b2] = lab2;

  const RAD = Math.PI / 180.0;
  const DEG = 180.0 / Math.PI;

  // Step 1
  const c1Ab = Math.sqrt(a1 * a1 + b1 * b1);
  const c2Ab = Math.sqrt(a2 * a2 + b2 * b2);
  const cAbMean = (c1Ab + c2Ab) / 2.0;

  const cAbMean7 = Math.pow(cAbMean, 7);
  const g = 0.5 * (1.0 - Math.sqrt(cAbMean7 / (cAbMean7 + Math.pow(25.0, 7))));

  const a1Prime = a1 * (1.0 + g);
  const a2Prime = a2 * (1.0 + g);

  const c1Prime = Math.sqrt(a1Prime * a1Prime + b1 * b1);
  const c2Prime = Math.sqrt(a2Prime * a2Prime + b2 * b2);

  let h1Prime = (Math.atan2(b1, a1Prime) * DEG) % 360.0;
  if (h1Prime < 0) h1Prime += 360.0;
  let h2Prime = (Math.atan2(b2, a2Prime) * DEG) % 360.0;
  if (h2Prime < 0) h2Prime += 360.0;

  // Step 2: Delta値
  const deltaLPrime = l2 - l1;
  const deltaCPrime = c2Prime - c1Prime;

  let deltaHPrime;
  if (c1Prime * c2Prime === 0.0) {
    deltaHPrime = 0.0;
  } else if (Math.abs(h2Prime - h1Prime) <= 180.0) {
    deltaHPrime = h2Prime - h1Prime;
  } else if (h2Prime - h1Prime > 180.0) {
    deltaHPrime = h2Prime - h1Prime - 360.0;
  } else {
    deltaHPrime = h2Prime - h1Prime + 360.0;
  }

  const deltaHPrimeBig = 2.0 * Math.sqrt(c1Prime * c2Prime) *
    Math.sin((deltaHPrime / 2.0) * RAD);

  // Step 3: CIEDE2000
  const lPrimeMean = (l1 + l2) / 2.0;
  const cPrimeMean = (c1Prime + c2Prime) / 2.0;

  let hPrimeMean;
  if (c1Prime * c2Prime === 0.0) {
    hPrimeMean = h1Prime + h2Prime;
  } else if (Math.abs(h1Prime - h2Prime) <= 180.0) {
    hPrimeMean = (h1Prime + h2Prime) / 2.0;
  } else if (h1Prime + h2Prime < 360.0) {
    hPrimeMean = (h1Prime + h2Prime + 360.0) / 2.0;
  } else {
    hPrimeMean = (h1Prime + h2Prime - 360.0) / 2.0;
  }

  const t = 1.0
    - 0.17 * Math.cos((hPrimeMean - 30.0) * RAD)
    + 0.24 * Math.cos((2.0 * hPrimeMean) * RAD)
    + 0.32 * Math.cos((3.0 * hPrimeMean + 6.0) * RAD)
    - 0.20 * Math.cos((4.0 * hPrimeMean - 63.0) * RAD);

  const lm50sq = (lPrimeMean - 50.0) * (lPrimeMean - 50.0);
  const sl = 1.0 + 0.015 * lm50sq / Math.sqrt(20.0 + lm50sq);
  const sc = 1.0 + 0.045 * cPrimeMean;
  const sh = 1.0 + 0.015 * cPrimeMean * t;

  const cPrimeMean7 = Math.pow(cPrimeMean, 7);
  const rc = 2.0 * Math.sqrt(cPrimeMean7 / (cPrimeMean7 + Math.pow(25.0, 7)));
  const deltaTheta = 30.0 * Math.exp(
    -Math.pow((hPrimeMean - 275.0) / 25.0, 2)
  );
  const rt = -Math.sin((2.0 * deltaTheta) * RAD) * rc;

  return Math.sqrt(
    Math.pow(deltaLPrime / sl, 2) +
    Math.pow(deltaCPrime / sc, 2) +
    Math.pow(deltaHPrimeBig / sh, 2) +
    rt * (deltaCPrime / sc) * (deltaHPrimeBig / sh)
  );
}

/**
 * パレットから最も近い色をCIEDE2000で検索。
 * @param {number} r - Red (0-255)
 * @param {number} g - Green (0-255)
 * @param {number} b - Blue (0-255)
 * @param {number[][]} palette - パレット色配列 [[r,g,b], ...]
 * @param {number} redPenalty - 赤ペナルティ係数 (0=無効)
 * @param {number} yellowPenalty - 黄ペナルティ係数 (0=無効)
 * @param {number} brightness - 正規化輝度 (0.0〜1.0)
 * @returns {number[]} 最近パレット色 [r, g, b]
 */
function findNearestColor(r, g, b, palette, redPenalty, yellowPenalty, brightness) {
  palette = palette || EINK_PALETTE;
  redPenalty = redPenalty || 0.0;
  yellowPenalty = yellowPenalty || 0.0;
  brightness = brightness || 0.0;

  const lab = rgbToLab(r, g, b);
  let bestColor = palette[0];
  let bestDist = Infinity;

  for (const p of palette) {
    let dist = ciede2000(lab, rgbToLab(p[0], p[1], p[2]));

    // 赤パレット色 (R>150, G<50, B<50) に明度ベースのペナルティ
    if (redPenalty > 0.0 && p[0] > 150 && p[1] < 50 && p[2] < 50) {
      dist += redPenalty * brightness;
    }
    // 黄パレット色 (R>200, G>200, B<50) に暗部ベースのペナルティ
    if (yellowPenalty > 0.0 && p[0] > 200 && p[1] > 200 && p[2] < 50) {
      dist += yellowPenalty * (1.0 - brightness);
    }

    if (dist < bestDist) {
      bestDist = dist;
      bestColor = p;
    }
  }

  return bestColor;
}

// --- sRGB → Linear LUT (256エントリ事前計算) ---

const SRGB_TO_LINEAR_LUT = new Float64Array(256);
for (let i = 0; i < 256; i++) {
  const v = i / 255.0;
  SRGB_TO_LINEAR_LUT[i] = v <= 0.04045 ? v / 12.92 : Math.pow((v + 0.055) / 1.055, 2.4);
}

// --- Lab変換定数 ---
const _LAB_DELTA = 6.0 / 29.0;
const _LAB_DELTA_SQ3 = _LAB_DELTA * _LAB_DELTA * _LAB_DELTA;
const _LAB_DELTA_SQ3_INV = 1.0 / (3.0 * _LAB_DELTA * _LAB_DELTA);
const _LAB_OFFSET = 4.0 / 29.0;
const _XN_INV = 1.0 / 0.95047;
const _ZN_INV = 1.0 / 1.08883;

/**
 * RGB(0-255) → Lab インライン高速変換。
 * rgbToLabと同じ結果だがLUTベースで高速。
 * @param {number} r - Red (0-255, integer)
 * @param {number} g - Green (0-255, integer)
 * @param {number} b - Blue (0-255, integer)
 * @returns {number[]} [L, a, b]
 */
function rgbToLabFast(r, g, b) {
  const rLin = SRGB_TO_LINEAR_LUT[r];
  const gLin = SRGB_TO_LINEAR_LUT[g];
  const bLin = SRGB_TO_LINEAR_LUT[b];

  const xr = (0.4124564 * rLin + 0.3575761 * gLin + 0.1804375 * bLin) * _XN_INV;
  const yr = 0.2126729 * rLin + 0.7151522 * gLin + 0.0721750 * bLin;
  const zr = (0.0193339 * rLin + 0.1191920 * gLin + 0.9503041 * bLin) * _ZN_INV;

  const fx = xr > _LAB_DELTA_SQ3 ? Math.pow(xr, 1.0 / 3.0) : xr * _LAB_DELTA_SQ3_INV + _LAB_OFFSET;
  const fy = yr > _LAB_DELTA_SQ3 ? Math.pow(yr, 1.0 / 3.0) : yr * _LAB_DELTA_SQ3_INV + _LAB_OFFSET;
  const fz = zr > _LAB_DELTA_SQ3 ? Math.pow(zr, 1.0 / 3.0) : zr * _LAB_DELTA_SQ3_INV + _LAB_OFFSET;

  return [116.0 * fy - 16.0, 500.0 * (fx - fy), 200.0 * (fy - fz)];
}


// --- LUT (ルックアップテーブル) ---

const LUT_STEP = 4;
const LUT_SIZE = 64; // 256 / LUT_STEP

/**
 * RGB→パレットインデックスの3D LUTを構築。
 * Lab Euclidean距離で最近色を決定。ペナルティなし。
 *
 * @param {number[][]} palette - パレット [[r,g,b], ...]
 * @returns {Uint8Array} 64³ のフラット配列。インデックス = (r>>2)*64*64 + (g>>2)*64 + (b>>2)
 */
function buildDitherLut(palette) {
  palette = palette || EINK_PALETTE;
  const palLab = palette.map(c => rgbToLabFast(c[0], c[1], c[2]));
  const nPal = palette.length;
  const lut = new Uint8Array(LUT_SIZE * LUT_SIZE * LUT_SIZE);
  const half = LUT_STEP >> 1; // 2

  for (let ri = 0; ri < LUT_SIZE; ri++) {
    const r = ri * LUT_STEP + half;
    for (let gi = 0; gi < LUT_SIZE; gi++) {
      const g = gi * LUT_STEP + half;
      for (let bi = 0; bi < LUT_SIZE; bi++) {
        const b = bi * LUT_STEP + half;
        const lab = rgbToLabFast(r, g, b);
        const pL = lab[0], pa = lab[1], pb = lab[2];

        let bestIdx = 0, bestDist = Infinity;
        for (let p = 0; p < nPal; p++) {
          const dL = pL - palLab[p][0];
          const da = pa - palLab[p][1];
          const db = pb - palLab[p][2];
          const dist = dL * dL + da * da + db * db;
          if (dist < bestDist) { bestDist = dist; bestIdx = p; }
        }
        lut[ri * LUT_SIZE * LUT_SIZE + gi * LUT_SIZE + bi] = bestIdx;
      }
    }
  }
  return lut;
}

/**
 * Lab Euclidean距離 + ペナルティで最近色インデックスを検索。
 * ペナルティあり時のディザリングループ用。
 *
 * @param {number} r - Red (0-255, integer)
 * @param {number} g - Green (0-255, integer)
 * @param {number} b - Blue (0-255, integer)
 * @param {number[][]} palette - パレット
 * @param {number[][]} palLab - パレットLab値 (事前計算済み)
 * @param {number} redPenalty - 赤ペナルティ係数
 * @param {number} yellowPenalty - 黄ペナルティ係数
 * @param {number} brightness - 正規化輝度 (0.0-1.0)
 * @returns {number} パレットインデックス
 */
function findNearestIndexLabEuclidean(r, g, b, palette, palLab, redPenalty, yellowPenalty, brightness) {
  const lab = rgbToLabFast(r, g, b);
  const pL = lab[0], pa = lab[1], pb = lab[2];
  let bestIdx = 0, bestDist = Infinity;

  for (let i = 0; i < palette.length; i++) {
    const dL = pL - palLab[i][0];
    const da = pa - palLab[i][1];
    const db = pb - palLab[i][2];
    let dist = Math.sqrt(dL * dL + da * da + db * db);

    const p = palette[i];
    if (redPenalty > 0 && p[0] > 150 && p[1] < 50 && p[2] < 50) {
      dist += redPenalty * brightness;
    }
    if (yellowPenalty > 0 && p[0] > 200 && p[1] > 200 && p[2] < 50) {
      dist += yellowPenalty * (1.0 - brightness);
    }

    if (dist < bestDist) { bestDist = dist; bestIdx = i; }
  }
  return bestIdx;
}

// Export for test and module usage
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    EINK_PALETTE, srgbToLinear, rgbToLab, rgbToLabFast, labToRgb, ciede2000,
    findNearestColor, findNearestIndexLabEuclidean,
    buildDitherLut, LUT_STEP, LUT_SIZE,
  };
}

File: src/epaper_palette_dither/domain/color.py
"""E-Ink 4色パレット定義と色距離計算。

Pure Pythonで実装（外部ライブラリ依存なし）。
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Sequence


@dataclass(frozen=True)
class RGB:
    """RGB色空間の色。各チャンネル 0-255。"""

    r: int
    g: int
    b: int

    def to_tuple(self) -> tuple[int, int, int]:
        return (self.r, self.g, self.b)


@dataclass(frozen=True)
class LAB:
    """CIE L*a*b* 色空間の色。"""

    l: float  # noqa: E741
    a: float
    b: float


# --- E-Ink 4色パレット ---

EINK_WHITE = RGB(255, 255, 255)
EINK_BLACK = RGB(0, 0, 0)
EINK_RED = RGB(200, 0, 0)
EINK_YELLOW = RGB(255, 255, 0)

EINK_PALETTE: tuple[RGB, ...] = (EINK_WHITE, EINK_BLACK, EINK_RED, EINK_YELLOW)

# --- 知覚パレット（e-paper実測値ベース） ---

EINK_WHITE_PERCEIVED = RGB(177, 175, 157)
EINK_BLACK_PERCEIVED = RGB(46, 38, 43)
EINK_RED_PERCEIVED = RGB(177, 51, 37)
EINK_YELLOW_PERCEIVED = RGB(198, 166, 26)

EINK_PALETTE_PERCEIVED: tuple[RGB, ...] = (
    EINK_WHITE_PERCEIVED, EINK_BLACK_PERCEIVED,
    EINK_RED_PERCEIVED, EINK_YELLOW_PERCEIVED,
)


# --- RGB → LAB 変換（Pure Python） ---


def _srgb_to_linear(c: int) -> float:
    """sRGBコンポーネント(0-255)をリニアRGBに変換。"""
    v = c / 255.0
    if v <= 0.04045:
        return v / 12.92
    return ((v + 0.055) / 1.055) ** 2.4


def _lab_f(t: float) -> float:
    """LAB変換の補助関数。"""
    delta = 6.0 / 29.0
    if t > delta**3:
        return t ** (1.0 / 3.0)
    return t / (3.0 * delta**2) + 4.0 / 29.0


def rgb_to_lab(color: RGB) -> LAB:
    """RGB色をCIE L*a*b*に変換。D65光源基準。"""
    # sRGB → リニアRGB
    r_lin = _srgb_to_linear(color.r)
    g_lin = _srgb_to_linear(color.g)
    b_lin = _srgb_to_linear(color.b)

    # リニアRGB → XYZ (D65)
    x = 0.4124564 * r_lin + 0.3575761 * g_lin + 0.1804375 * b_lin
    y = 0.2126729 * r_lin + 0.7151522 * g_lin + 0.0721750 * b_lin
    z = 0.0193339 * r_lin + 0.1191920 * g_lin + 0.9503041 * b_lin

    # D65 白色点
    xn, yn, zn = 0.95047, 1.00000, 1.08883

    # XYZ → L*a*b*
    fx = _lab_f(x / xn)
    fy = _lab_f(y / yn)
    fz = _lab_f(z / zn)

    l_star = 116.0 * fy - 16.0
    a_star = 500.0 * (fx - fy)
    b_star = 200.0 * (fy - fz)

    return LAB(l_star, a_star, b_star)


# --- 色距離計算（CIEDE2000） ---


def ciede2000(lab1: LAB, lab2: LAB) -> float:
    """CIEDE2000色差を計算。

    参考: "The CIEDE2000 Color-Difference Formula" (Sharma et al., 2005)
    """
    l1, a1, b1 = lab1.l, lab1.a, lab1.b
    l2, a2, b2 = lab2.l, lab2.a, lab2.b

    # Step 1: 計算
    c1_ab = math.sqrt(a1**2 + b1**2)
    c2_ab = math.sqrt(a2**2 + b2**2)
    c_ab_mean = (c1_ab + c2_ab) / 2.0

    c_ab_mean_7 = c_ab_mean**7
    g = 0.5 * (1.0 - math.sqrt(c_ab_mean_7 / (c_ab_mean_7 + 25.0**7)))

    a1_prime = a1 * (1.0 + g)
    a2_prime = a2 * (1.0 + g)

    c1_prime = math.sqrt(a1_prime**2 + b1**2)
    c2_prime = math.sqrt(a2_prime**2 + b2**2)

    h1_prime = math.degrees(math.atan2(b1, a1_prime)) % 360.0
    h2_prime = math.degrees(math.atan2(b2, a2_prime)) % 360.0

    # Step 2: Delta値
    delta_l_prime = l2 - l1
    delta_c_prime = c2_prime - c1_prime

    if c1_prime * c2_prime == 0.0:
        delta_h_prime = 0.0
    elif abs(h2_prime - h1_prime) <= 180.0:
        delta_h_prime = h2_prime - h1_prime
    elif h2_prime - h1_prime > 180.0:
        delta_h_prime = h2_prime - h1_prime - 360.0
    else:
        delta_h_prime = h2_prime - h1_prime + 360.0

    delta_H_prime = 2.0 * math.sqrt(c1_prime * c2_prime) * math.sin(
        math.radians(delta_h_prime / 2.0)
    )

    # Step 3: CIEDE2000
    l_prime_mean = (l1 + l2) / 2.0
    c_prime_mean = (c1_prime + c2_prime) / 2.0

    if c1_prime * c2_prime == 0.0:
        h_prime_mean = h1_prime + h2_prime
    elif abs(h1_prime - h2_prime) <= 180.0:
        h_prime_mean = (h1_prime + h2_prime) / 2.0
    elif h1_prime + h2_prime < 360.0:
        h_prime_mean = (h1_prime + h2_prime + 360.0) / 2.0
    else:
        h_prime_mean = (h1_prime + h2_prime - 360.0) / 2.0

    t = (
        1.0
        - 0.17 * math.cos(math.radians(h_prime_mean - 30.0))
        + 0.24 * math.cos(math.radians(2.0 * h_prime_mean))
        + 0.32 * math.cos(math.radians(3.0 * h_prime_mean + 6.0))
        - 0.20 * math.cos(math.radians(4.0 * h_prime_mean - 63.0))
    )

    sl = 1.0 + 0.015 * (l_prime_mean - 50.0) ** 2 / math.sqrt(
        20.0 + (l_prime_mean - 50.0) ** 2
    )
    sc = 1.0 + 0.045 * c_prime_mean
    sh = 1.0 + 0.015 * c_prime_mean * t

    c_prime_mean_7 = c_prime_mean**7
    rc = 2.0 * math.sqrt(c_prime_mean_7 / (c_prime_mean_7 + 25.0**7))
    delta_theta = 30.0 * math.exp(
        -(((h_prime_mean - 275.0) / 25.0) ** 2)
    )
    rt = -math.sin(math.radians(2.0 * delta_theta)) * rc

    return math.sqrt(
        (delta_l_prime / sl) ** 2
        + (delta_c_prime / sc) ** 2
        + (delta_H_prime / sh) ** 2
        + rt * (delta_c_prime / sc) * (delta_H_prime / sh)
    )


def find_nearest_color_index(
    color: RGB,
    palette: Sequence[RGB] = EINK_PALETTE,
    red_penalty: float = 0.0,
    yellow_penalty: float = 0.0,
    brightness: float = 0.0,
    perceived_palette: Sequence[RGB] | None = None,
) -> int:
    """パレットから最も近い色のインデックスをCIEDE2000で検索。

    Args:
        color: 検索対象の色
        palette: 出力用カラーパレット
        red_penalty: 赤パレット色へのペナルティ係数 (0=無効)
        yellow_penalty: 黄パレット色へのペナルティ係数 (0=無効)
        brightness: 正規化輝度 (0.0〜1.0)。ペナルティと組み合わせて使用
        perceived_palette: 知覚パレット。指定時はCIEDE2000距離を知覚値で計算

    Returns:
        最近傍色のインデックス
    """
    lab = rgb_to_lab(color)
    dist_palette = perceived_palette if perceived_palette is not None else palette
    best_idx = 0
    best_dist = float("inf")

    for i, (p, dp) in enumerate(zip(palette, dist_palette)):
        dist = ciede2000(lab, rgb_to_lab(dp))
        # ペナルティ判定は常に出力パレットで行う
        if red_penalty > 0.0 and p.r > 150 and p.g < 50 and p.b < 50:
            dist += red_penalty * brightness
        if yellow_penalty > 0.0 and p.r > 200 and p.g > 200 and p.b < 50:
            dist += yellow_penalty * (1.0 - brightness)
        if dist < best_dist:
            best_dist = dist
            best_idx = i

    return best_idx


def find_nearest_color(
    color: RGB,
    palette: Sequence[RGB] = EINK_PALETTE,
    red_penalty: float = 0.0,
    yellow_penalty: float = 0.0,
    brightness: float = 0.0,
    perceived_palette: Sequence[RGB] | None = None,
) -> RGB:
    """パレットから最も近い色をCIEDE2000で検索。

    Args:
        color: 検索対象の色
        palette: 出力用カラーパレット
        red_penalty: 赤パレット色へのペナルティ係数 (0=無効)
        yellow_penalty: 黄パレット色へのペナルティ係数 (0=無効)
        brightness: 正規化輝度 (0.0〜1.0)。ペナルティと組み合わせて使用
        perceived_palette: 知覚パレット。指定時はCIEDE2000距離を知覚値で計算
    """
    idx = find_nearest_color_index(
        color, palette, red_penalty, yellow_penalty, brightness, perceived_palette,
    )
    return palette[idx]
